### PR TITLE
polish up eth_bots

### PR DESCRIPTION
**Pull request by dpaiton**
**DONE:**

- `elfpy/agents/policies/random_agent.py` 
    - 204: # FIXME: Close policies when out of money & if some are open

    
- `elfpy/markets/hyperdrive/hyperdrive_assets.py`
    - 5: # FIXME: Move these into hyperdrive_interface/

- `examples/eth_bots/execute_agent_trades.py`
    - 148: # FIXME: Switch this to a match statement

- `examples/eth_bots/setup_agents.py`
    - 32: # FIXME: add better comments explaining this process
    
- `examples/eth_bots/main.py`
    - 25: # FIXME: move this out of this file (into `elfpy/bots/register_username_server.py`?
    - 34: # FIXME: Move much of this out of main
    - 37: # FIXME: all contract initialization should get encapsulated into something like 'setup_experiment()'
    - 81: # FIXME: move into get_agent_accounts
    - 87: # FIXME: encapulate trade loop to another function.  At most should be:

**NOT GOING TO DO THIS PR:**

- `examples/eth_bots/trade_loop.py`
    - 67: # TODO: crash reporting

- `elfpy/agents/policies/base.py`
    - 25: # TODO: Check that multiple agent.rng derefs to the same rng object

- `elfpy/bots/bot_info.py`
    - 14:  # TODO: Rename to `AgentConfig` when we remove evm_bots

- `examples/eth_bots/setup_agents.py`
    - 21: # TODO: raise issue on failure by looking at `rpc_response`, `tx_receipt` returned from function

- `examples/eth_bots/main.py`
    - 6: # TODO: Move configs into a dedicated config folder with the other elfpy configs

- `examples/eth_bots/execute_agent_trades.py`
    - 71: # TODO: raise issue on failure by looking at `tx_receipt` returned from function

- `examples/eth_bots/config.py`
    - 61: # TODO: Clean up variables that are only used by evm_bots & apeworx
    - 70: # TODO: should just use hostname, artifacts_port, etc.

- `elfpy/eth/transactions.py`
    - 101: # TODO: Should not wait for approval since the block might take a bit to tick up. This also means we need to parse the tx_hash, not the tx_receipt.
    
- `elfpy/hyperdrive_interface/hyperdrive_interface.py`
    - 246: # TODO: Should get the anvil increment time
    
- `elfpy/agents/policies/smart_long.py`
    - 71: # TODO: should we not wait until maturity? they never close

### CODE REVIEW

File: examples/eth_bots/config.py [hunk 1]
```diff
@@ -58,17 +58,26 @@
...
-    username="changeme",
+    # TODO: should just use
+    # hostname = http://localhost
+    # artifacts_port = 80
+    # rpc_port = 8545
+    # user_registry_port = 5002
+    artifacts_url="http://localhost:80",
+    rpc_url="http://localhost:8545",
     username_register_url="http://localhost:5002",
+    # artifacts_url="http://ec2-3-13-94-236.us-east-2.compute.amazonaws.com:80",
```
Comment (slundqui) [inline examples/eth_bots/config.py:79]: nit, I'd say lets leave out the actual url endpoint from the repo. I guess now it's committed though it's in the history forever...

**Review by slundqui — APPROVED** [review]
Looks good, most of it looks like it's cleanup

File: examples/eth_bots/setup_experiment.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+from elfpy.eth.accounts import EthAccount
+from elfpy.utils import logs
+from examples.eth_bots.setup_agents import get_agent_accounts
+
+
+def register_username(register_url: str, wallet_addrs: list[str], username: str) -> None:
+    """Connects to the register user flask server via post request and registeres the username"""
+    json_data = {"wallet_addrs": wallet_addrs, "username": username}
+    result = requests.post(register_url + "/register_bots", json=json_data, timeout=3)
+    if result.status_code != 200:
```
Comment (wakamex) [inline examples/eth_bots/setup_experiment.py:22]:
```suggestion
    if result.status_code != HTTPStatus.OK:
```

File: examples/eth_bots/setup_experiment.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+"""Setup helper function for running eth bot experiments"""
+from __future__ import annotations
+
+import os
+
+import numpy as np
+import requests
+from web3 import Web3
```
Comment (wakamex) [inline examples/eth_bots/setup_experiment.py:8]:
```suggestion
from http import HTTPStatus
from web3 import Web3
```

File: examples/eth_bots/config.py (hunk 1 above)
Comment (sentilesdal) [reply examples/eth_bots/config.py]: I also would have said this. hmmm.  we _can_ overwrite this.  the risk is pretty low in this case so we're lucky.  remember these are totally unsecure ports.

**Review by wakamex — COMMENTED** [review]
avoid [magic value comparison](https://beta.ruff.rs/docs/rules/magic-value-comparison/)
ruff doesn't like it: `examples/eth_bots/main.py:31:30: PLR2004 Magic value used in comparison, consider replacing 200 with a constant variable`

> Why is this bad?
> The use of "magic" values can make code harder to read and maintain, as readers will have to infer the meaning of the value from the context. Such values are discouraged by [PEP 8](https://peps.python.org/pep-0008/#constants).

File: examples/eth_bots/setup_agents.py [hunk 4]
```diff
@@ -4,25 +4,60 @@
...
 from elfpy import eth
 from elfpy.agents.agent import Agent
-from elfpy.bots.bot_info import BotInfo
+from elfpy.bots import BotInfo, EnvironmentConfig
 from elfpy.eth.accounts import EthAccount
 
 
+def register_username(register_url: str, wallet_addrs: list[str], username: str) -> None:
+    """Connects to the register user flask server via post request and registeres the username"""
+    json_data = {"wallet_addrs": wallet_addrs, "username": username}
```
Comment (sentilesdal) [inline examples/eth_bots/setup_agents.py]: TODO: use the json schema from the server.

File: examples/eth_bots/config.py (hunk 1 above)
Comment (dpaiton) [reply examples/eth_bots/config.py:79]: no it'll be gone on squash. I'll remove it.

File: elfpy/eth/rpc_interface.py [hunk 5]
```diff
@@ -34,3 +34,28 @@ def get_account_balance(web3: Web3, account_address: str) -> int | None:
     if hex_result is not None:
         return int(hex_result, base=16)  # cast hex to int
     return None
+
+
+def get_wait_for_new_block(web3: Web3) -> bool:
```
Comment (sentilesdal) [inline elfpy/eth/rpc_interface.py:39]: sorry i misled you.  this is not really a generalized function that belongs in rpc_interface, its pretty opinionated.  should just be a private helper function in the file its being used.

File: examples/eth_bots/execute_agent_trades.py [hunk 6]
```diff
@@ -137,124 +137,124 @@ def execute_agent_trades(
...
-                    balance=Quantity(
-                        amount=trade_result.base_amount,
-                        unit=TokenType.BASE,
-                    ),
-                    lp_tokens=-trade_result.lp_amount,
-                    withdraw_shares=trade_result.withdrawal_share_amount,
-                )
-            else:
-                raise NotImplementedError(f"{trade_object.trade.action_type} is not implemented.")
+            match trade_object.trade.action_type:
```
Comment (sentilesdal) [inline examples/eth_bots/execute_agent_trades.py]: why we're at it, lets make this an exhaustive match!

```python
from typing import NoReturn


class Things(Enum):
    ONE = "one"
    TWO = "two"


def assert_never(x: NoReturn) -> NoReturn:
    assert False, "Unhandled type: {}".format(type(x).__name__)


def handle_not_exhaustive(x: Things) -> None:
    match x:
        case x.ONE:
            print(f"Handled {x}")

        case _:
            # Error: Argument of type "Literal[Things.TWO]" cannot be assigned to parameter "x" of
            # type "NoReturn" in function "assert_never"
            assert_never(x)


def handle_exhaustive(x: Things) -> None:
    match x:
        case x.ONE:
            print(f"Handled {x}")

        case x.TWO:
            print(f"Handled {x}")

        case _:
            assert_never(x)  # No error
```

File: examples/eth_bots/execute_agent_trades.py [hunk 7]
```diff
@@ -128,133 +129,171 @@ def execute_agent_trades(
...
+
+
+def match_contract_call_to_trade(
+    web3: Web3,
+    hyperdrive_contract: Contract,
+    hyperdrive_market: HyperdriveMarket,
+    account: EthAccount,
+    trade_object: elftypes.Trade,
+) -> WalletDeltas:
+    """Switch statement that executes the smart contract trade based on the provided type.
```
Comment (sentilesdal) [inline examples/eth_bots/execute_agent_trades.py:148]: Ahem, 'match' statement ;)

Comment (sentilesdal) [reply examples/eth_bots/execute_agent_trades.py:148]: this is nice I like isolating this logic.

**Review by sentilesdal — APPROVED** [review]
looks good!  you know dis but there's like 5 prs in here.

File: examples/eth_bots/setup_experiment.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
+"""Setup helper function for running eth bot experiments"""
```
Comment (wakamex) [inline examples/eth_bots/setup_experiment.py:1]:
```suggestion
"""Setup helper function for running eth bot experiments."""
```

File: examples/eth_bots/setup_experiment.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+from elfpy import eth, hyperdrive_interface
+from elfpy.bots import DEFAULT_USERNAME, BotInfo, EnvironmentConfig
+from elfpy.eth.accounts import EthAccount
+from elfpy.utils import logs
+from examples.eth_bots.setup_agents import get_agent_accounts
+
+
+def register_username(register_url: str, wallet_addrs: list[str], username: str) -> None:
+    """Connects to the register user flask server via post request and registeres the username"""
```
Comment (wakamex) [inline examples/eth_bots/setup_experiment.py:20]:
```suggestion
    """Registers the username with the flask server."""
```

**Pull request merged**